src/goto-instrument/contracts/contracts.cpp

```diff
@@ -387,10 +398,19 @@ void code_contractst::replace_old_parameter(
     replace_old_parameter(op, parameter2history, location, mode, history);
   }
 
-  if(expr.id() == ID_old)
+  if(expr.id() == ID_old ||
+     expr.id() == ID_loop_entry)
   {
-    DATA_INVARIANT(
+    if (expr.id() == ID_old)
+    {
+      DATA_INVARIANT(
       expr.operands().size() == 1, CPROVER_PREFIX "old must have one operand");
+    }
+    if (expr.id() == ID_loop_entry)
+    {
+      DATA_INVARIANT(
+      expr.operands().size() == 1, CPROVER_PREFIX "loop_entry must have one operand");
+    }
```

Comment:
@SaswatPadhi if we add info of the source location, we could perhaps avoid code duplication and keep a single error message for both history variables. What do you think? Only a suggestion.

Reply:
This is an internal invariant (`DATA_INVARIANT`) failure, not a user-facing error message; so I don't think source location would make much sense.

But looking at this code now, I don't think this entire thing is necessary at all. You can remove lines 404 - 414.

The `to_old_expr` call @ line 415 would call `validate_expr` for the base class `unary_expr`, which would anyway check the number of operands internally:

https://github.com/diffblue/cbmc/blob/5b8e8fa6fc8eb6b241aaca38e4e706568a7ac158/src/util/std_expr.h#L317-L320.

If we want to show user-facing errors, that is already done within the parser.